### PR TITLE
Ventcrawl port from Bay12

### PR DESCRIPTION
**Pull request by Bedshaped**
Original code from vg, port credit goes to Zuhayr.

DNM, I still need to iron out some bugs and tighten the code on level 219.

First test

### CODE REVIEW

File: code/modules/ventcrawl/ventcrawl.dm [hunk 1]
```diff
@@ -0,0 +1,210 @@
+var/list/ventcrawl_machinery = list(
+	/obj/machinery/atmospherics/unary/vent_pump,
+	/obj/machinery/atmospherics/unary/vent_scrubber
+	)
+
+// Vent crawling whitelisted items, whoo
+/mob/living/var/list/can_enter_vent_with = list(
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: I wonder if it'd be easier to use an item's `w_class` instead, and to permit only values of 1 and maybe 2?

File: code/modules/mob/living/carbon/human/life.dm [hunk 2]
```diff
@@ -1282,9 +1282,10 @@
 					*/
 
 		else
-			sight = species.vision_flags
-			see_in_dark = species.darksight
-			see_invisible = see_in_dark>2 ? SEE_INVISIBLE_LEVEL_ONE : SEE_INVISIBLE_LIVING
+			if(is_ventcrawling == 0) // Stops sight returning to normal if inside a vent
+				sight = species.vision_flags
+				see_in_dark = species.darksight
+				see_invisible = see_in_dark>2 ? SEE_INVISIBLE_LEVEL_ONE : SEE_INVISIBLE_LIVING
```
Comment (Bedshaped) [inline code/modules/mob/living/carbon/human/life.dm]: I think I need to put this higher in the tree so it doesn't just work with monkeys

File: code/modules/ventcrawl/ventcrawl.dm [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		if(istype(carried_item, can_enter_vent_with))
+			return get_inventory_slot(carried_item) == 0
+	return 0
+
+/mob/living/carbon/is_allowed_vent_crawl_item(var/obj/item/carried_item)
+	if(carried_item in internal_organs)
+		return 1
+	return ..()
+
+/mob/living/carbon/human/is_allowed_vent_crawl_item(var/obj/item/carried_item)
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: This is unnecessary. If the parent (on line 42) is defined the same way, then there's no point in redefining the child as well. It inherits fully already.

File: code/modules/ventcrawl/ventcrawl.dm [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	return ..()
+
+/mob/living/proc/ventcrawl_carry()
+	for(var/atom/A in src.contents)
+		if(!is_allowed_vent_crawl_item(A))
+			src << "<span class='warning'>You can't be carrying that when vent crawling!</span>"
+			return 0
+	return 1
+
+/mob/living/AltClickOn(var/atom/A)
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: Is there a way to flip this around? Meaning, is there a proc which is fired on an object when someone alt-clicks on it? That way, you would define for both scrubbers and vents. Though, hrm, that'd be duplicate definitions, effectively. Hrm.

File: code/modules/ventcrawl/ventcrawl.dm [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+/mob/living/var/list/icon/pipes_shown = list()
+/mob/living/var/last_played_vent
+/mob/living/var/is_ventcrawling = 0
+/mob/var/next_play_vent = 0
+
+/mob/living/proc/can_ventcrawl()
+	return 0
+
+/mob/living/Login()
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: I wonder if this should be taken back to the original definition of mob/living/Login. Unless the proc is already parsed out?

File: code/modules/ventcrawl/ventcrawl.dm [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	return issmall(src)
+
+/mob/proc/start_ventcrawl()
+	var/atom/pipe
+	var/list/pipes = list()
+	for(var/obj/machinery/atmospherics/unary/U in range(1))
+		if(is_type_in_list(U,ventcrawl_machinery) && Adjacent(U))
+			pipes |= U
+	if(!pipes || !pipes.len)
+		usr << "<span class='notice'>There are no pipes that you can ventcrawl into within range!"
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: Unclosed span.

File: code/modules/ventcrawl/ventcrawl.dm [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+						vent_found = machine
+
+					if(!vent_found || !vent_found.can_crawl_through())
+						vent_found = null
+
+					if(vent_found)
+						break
+
+			if(istype(vent_found, /obj/machinery/atmospherics/unary/vent_pump/) && vent_found:is_welded()) // welded check
+				usr << "<span class='warning'>You can't crawl into a welded vent!</span>"
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: Change this to `src`. Same for every other mention of `usr` in this proc.

File: code/modules/ventcrawl/ventcrawl_atmospherics.dm [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+/obj/machinery/atmospherics/proc/ventcrawl_to(var/mob/living/user, var/obj/machinery/atmospherics/target_move, var/direction)
+	if(target_move)
+		if(is_type_in_list(target_move, ventcrawl_machinery) && target_move.can_crawl_through())
+			if(istype(target_move, /obj/machinery/atmospherics/unary/vent_pump/) && target_move:is_welded())
+				usr.visible_message("<span class='warning'You hear something banging on \the [target_move.name]!</span>", "<span class='notice'>You can't escape from a welded vent.</span>")
+			else
+				user.remove_ventcrawl()
+				user.forceMove(target_move.loc) //handles entering and so on
+				user.sight &= ~(SEE_TURFS|BLIND)
+				usr.visible_message("<span class='warning'>You hear something squeezing through the ducts.</span>", "You climb out the ventilation system.")
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl_atmospherics.dm]: This should use `user` instead of `usr`. Same for every other use of `usr` in this proc.

**Comment by skull132** [comment]
One general note: please standardize the level at which you implement the vent crawl procs and vars for mobs. Some procs and vars are at mob/living level, others are at mob level.

File: code/modules/ventcrawl/ventcrawl.dm [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+	/obj/item/device/radio/borg,
+	/obj/item/weapon/holder,
+	/obj/machinery/camera,
+	/mob/living/simple_animal/borer
+	)
+
+/mob/living/var/list/icon/pipes_shown = list()
+/mob/living/var/last_played_vent
+/mob/living/var/is_ventcrawling = 0
+/mob/var/next_play_vent = 0
```
Comment (skull132) [inline code/modules/ventcrawl/ventcrawl.dm]: Still mob level, should be `mob/living/var/next_play_vent = 0`.

**Pull request merged**